lib/open_project/scm/adapters/git.rb

```diff
@@ -91,6 +114,82 @@ def check_availability!
           raise Exceptions::ScmUnavailable
         end
 
+        def checkout_repository!
+          Rails.logger.info "Checking out #{checkout_uri} to #{checkout_path}"
+
+          FileUtils.mkdir_p checkout_path.parent
+          %x(git clone #{checkout_uri} #{checkout_path})
+
+          if $? != 0
+            raise Exceptions::CommandFailed("Failed to clone #{checkout_uri} to #{checkout_path}")
+          end
+        end
+
+        def update_repository!
+          Rails.logger.debug "Fetching latest commits for #{checkout_path}"
+
+          Dir.chdir checkout_path do
+            fetch_all
+
+            remote_branches = self.remote_branches
+            local_branches = self.local_branches
+
+            remote_branches.each do |remote_branch|
+              local_branch = remote_branch.sub /^origin\//, ""
+
+              if !local_branches.include?(local_branch)
+                track_branch! local_branch, remote_branch
+              else
+                update_branch! local_branch, remote_branch
+              end
+            end
+          end
+        end
+
+        def fetch_all
+          %x(#{client_command} fetch --all)
+
+          raise Exceptions::CommandFailed.new("git", "Failed to fetch latest commits") if $? != 0
```

Comment:
If you use `capture_out`, it will automatically raise an exception if the exit code is not zero.